C/C++ codegen bug with ``deepCopy`` magic

Compiled with ``--threads:on``:

```nim
import threadpool

type
  Test = ref object
    x: int

proc `$`(v: Test): string = v[].`$`

let x = Test(x: 1)

proc test[T](p: pointer) {.thread.} =
  var v: T
  # This leads to codegen error:
  deepCopy(v, cast[T](p))
  # And this works:
  # let pv = cast[T](p)
  # deepCopy(v, pv)
  echo "THREAD: ", v

spawn test[type(x)](cast[pointer](x))

sync()

echo "Done"
```

produces:

```
C: nimtests_deepcopy_test
Error: execution of an external compiler program 'gcc -c  -w -pthread  -I/home/vega/work/oss/Nim/lib -o /home/vega/work/oss/nim-projects/nimtests/lang/threading/nimcache/nimtests_deepcopy_test.o /home/vega/work/oss/nim-projects/nimtests/lang/threading/nimcache/nimtests_deepcopy_test.c' failed with exit code: 1

/home/vega/work/oss/nim-projects/nimtests/lang/threading/nimcache/nimtests_deepcopy_test.c: In function ‘test_pwiQdYJVKZVMoYZXUzOrZQ’:
/home/vega/work/oss/nim-projects/nimtests/lang/threading/nimcache/nimtests_deepcopy_test.c:426:40: error: lvalue required as unary ‘&’ operand
  genericDeepCopy((void*)(&v0), (void*)(&((TestHEX3Aobjecttype_zhD1YC1TxBDGyDzjJNzyhg*) (p0))), (&NTI_umpyhpeDIUVZB5AoYBkndg));
```

--- a/compiler/ccgexprs.nim
+++ b/compiler/ccgexprs.nim
@@ -363,19 +363,28 @@ proc genAssignment(p: BProc, dest, src: TLoc, flags: TAssignmentFlags) =
             rope p.currLineInfo.safeLineNm)
 
 proc genDeepCopy(p: BProc; dest, src: TLoc) =
+  template addrLocOrTemp(a: TLoc): Rope =
+    if a.k == locExpr:
+      var tmp: TLoc
+      getTemp(p, a.t, tmp)
+      genAssignment(p, tmp, a, {})
+      addrLoc(tmp)
+    else:
+      addrLoc(a)
+
   var ty = skipTypes(dest.t, abstractVarRange)
   case ty.kind
   of tyPtr, tyRef, tyProc, tyTuple, tyObject, tyArray:
     # XXX optimize this
     linefmt(p, cpsStmts, "#genericDeepCopy((void*)$1, (void*)$2, $3);$n",
-            addrLoc(dest), addrLoc(src), genTypeInfo(p.module, dest.t))
+            addrLoc(dest), addrLocOrTemp(src), genTypeInfo(p.module, dest.t))
   of tySequence, tyString:
     linefmt(p, cpsStmts, "#genericSeqDeepCopy($1, $2, $3);$n",
             addrLoc(dest), rdLoc(src), genTypeInfo(p.module, dest.t))
   of tyOpenArray, tyVarargs:
     linefmt(p, cpsStmts,
          "#genericDeepCopyOpenArray((void*)$1, (void*)$2, $1Len0, $3);$n",
-         addrLoc(dest), addrLoc(src), genTypeInfo(p.module, dest.t))
+         addrLoc(dest), addrLocOrTemp(src), genTypeInfo(p.module, dest.t))
   of tySet:
     if mapType(ty) == ctArray:
       useStringh(p.module)

--- a/tests/ccgbugs/tdeepcopy_addr_rval.nim
+++ b/tests/ccgbugs/tdeepcopy_addr_rval.nim
@@ -1,0 +1,16 @@
+discard """
+  output: "3"
+"""
+
+# issue 5166
+
+type
+  Test = ref object
+    x: int
+
+let x = Test(x: 3)
+let p = cast[pointer](x)
+
+var v: Test
+deepCopy(v, cast[Test](p))
+echo v.x

